crates/walrus-service/src/node/checkpoint.rs

```diff
@@ -0,0 +1,820 @@
+    fn purge_old_checkpoints(checkpoint_dir: &Path, max_checkpoints: usize) {
+        if max_checkpoints == 0 {
+            return;
+        }
+
+        let Ok(mut engine) = Self::create_backup_engine(checkpoint_dir) else {
+            tracing::error!("Failed to create backup engine");
+            return;
+        };
+        let result = engine
+            .purge_old_backups(max_checkpoints)
+            .map_err(|e| CheckpointError::Other(anyhow::anyhow!("Purge error: {}", e)));
+
+        match result {
+            Ok(_) => tracing::info!(
+                "purged old checkpoints, keeping {} most recent",
+                max_checkpoints
+            ),
+            Err(e) => tracing::error!(?e, "Failed to purge old checkpoints"),
+        }
+    }
+
+    /// Restore from the most recent backup.
+    pub async fn restore_latest(
+        checkpoint_dir: &Path,
+        db_path: &Path,
+        wal_dir: Option<&Path>,
+    ) -> Result<(), CheckpointError> {
+        // Create a fresh BackupEngine for this operation.
+        let mut engine = Self::create_backup_engine(checkpoint_dir)?;
+
+        let restore_opts = RestoreOptions::default();
+        let wal_path = wal_dir.unwrap_or(db_path);
+
+        tracing::info!("restoring database from latest backup");
+        engine
+            .restore_from_latest_backup(db_path, wal_path, &restore_opts)
+            .map_err(|e| CheckpointError::Other(anyhow::anyhow!("restore error: {}", e)))?;
+
+        tracing::info!("database restored successfully");
+        Ok(())
+    }
+
+    fn create_backup_impl(
+        db: &Arc<RocksDB>,
+        checkpoint_dir: &Path,
+        flush_before_backup: bool,
+    ) -> Result<(), CheckpointError> {
+        let mut engine = Self::create_backup_engine(checkpoint_dir)?;
+
+        let start_time = Instant::now();
+        tracing::info!(
+            checkpoint_dir = ?checkpoint_dir,
+            "start creating RocksDB backup"
+        );
+
+        let db_ref = &db.underlying;
+        let backup_result = engine
+            .create_new_backup_flush(db_ref, flush_before_backup)
+            .map_err(|e| CheckpointError::Other(anyhow::anyhow!("Backup error: {}", e)));
+
+        let duration = start_time.elapsed();
+
+        match backup_result {
+            Ok(_) => {
+                tracing::info!(
+                    duration = ?duration,
+                    "rocksDB backup created successfully"
+                );
+                Ok(())
+            }
+            Err(e) => {
+                tracing::error!(
+                    ?e,
+                    duration = ?duration,
+                    "rocksDB backup failed",
+                );
+                Err(e)
+            }
+        }
+    }
+
+    /// Create the checkpoint directory if it doesn't exist.
+    fn create_checkpoint_dir_if_not_exists(checkpoint_dir: &Path) -> Result<(), CheckpointError> {
+        match checkpoint_dir.try_exists() {
```

Comment:
You can just call create_dir_all, don't check whether it already exists. It does this already and succeeds if the dir already exists.